geth/jail/handlers.go

```diff
@@ -4,134 +4,209 @@ import (
 	}
 
-	// register send handler
-	if err = registerHandler("send", makeSendHandler(jail, cell)); err != nil {
+	if err := cell.Set("jeth", jeth); err != nil {
 		return err
 	}
 
-	// register sendAsync handler
-	if err = registerHandler("sendAsync", makeAsyncSendHandler(jail, cell)); err != nil {
-		return err
-	}
+	return nil
+}
 
-	// register isConnected handler
-	if err = registerHandler("isConnected", makeJethIsConnectedHandler(jail, cell)); err != nil {
-		return err
+// registerStatusSignals creates an object called "statusSignals".
+// TODO(adam): describe what it is and when it's used.
+func registerStatusSignals(jail *Jail, cell *Cell) error {
+	statusSignals := map[string]interface{}{
+		"sendSignal": createSendSignalHandler(jail, cell),
 	}
 
-	// register sendMessage/showSuggestions handlers
-	if err = cell.Set("statusSignals", struct{}{}); err != nil {
-		return err
-	}
-	statusSignals, err := cell.Get("statusSignals")
-	if err != nil {
-		return err
-	}
-	registerHandler = statusSignals.Object().Set
-	if err = registerHandler("sendSignal", makeSignalHandler(chatID)); err != nil {
+	if err := cell.Set("statusSignals", statusSignals); err != nil {
 		return err
 	}
 
 	return nil
 }
 
-// makeAsyncSendHandler returns jeth.sendAsync() handler.
-func makeAsyncSendHandler(jail *Jail, cellInt common.JailCell) func(call otto.FunctionCall) otto.Value {
-	// FIXME(tiabc): Get rid of this.
-	cell := cellInt.(*Cell)
+// createSendHandler returns jeth.send() and jeth.sendAsync() handler
+func createSendHandler(jail *Jail, cell *Cell) func(call otto.FunctionCall) otto.Value {
 	return func(call otto.FunctionCall) otto.Value {
-		go func() {
-			response := jail.Send(call)
+		// As it's a sync call, it's called already from a thread-safe context,
+		// thus using otto.Otto directly. Otherwise, it would try to acquire a lock again
+		// and result in a deadlock.
+		vm := cell.VM.UnsafeVM()
+
+		request, err := vm.Call("JSON.stringify", nil, call.Argument(0))
+		if err != nil {
+			throwJSError(err)
+		}
+
+		response, err := jail.sendRPCCall(cell, request.String())
+		if err != nil {
+			throwJSError(err)
+		}
+
+		value, err := vm.ToValue(response)
+		if err != nil {
+			throwJSError(err)
+		}
+
+		return value
+	}
+}
 
+// createSendAsyncHandler returns jeth.sendAsync() handler.
+func createSendAsyncHandler(jail *Jail, cell *Cell) func(call otto.FunctionCall) otto.Value {
+	return func(call otto.FunctionCall) otto.Value {
+		go func() {
+			// As it's an async call, it's not called from a thread-safe context,
+			// thus using vm.VM.
+			vm := cell.VM
 			callback := call.Argument(1)
-			if callback.Class() == "Function" {
-				// run callback asyncronously with args (error, response)
-				err := otto.NullValue()
-				cell.CallAsync(callback, err, response)
+			isFunction := callback.Class() == "Function"
+
+			request, err := vm.Call("JSON.stringify", nil, call.Argument(0))
+			if err != nil && isFunction {
```

Comment:
Why check for `isFunction` here? `isFunction` actually should be named `hasCallback`.
So, error in converting call from JSON to otto should be a failure regardless if this send call has callback or not (as I understand `sendAsync` without callback is a perfectly valid case)

Reply:
Because otherwise I will pass a non-function to `CallAsync` and we should avoid that, right?

> isFunction actually should be named hasCallback

It sounds good.

> So, error in converting call from JSON to otto should be a failure regardless if this send call has callback or not

Correct. I should do it before a goroutine then, otherwise, there is no chance to report it back.

> as I understand sendAsync without callback is a perfectly valid case)

Correct.